src/use_file.rs

```diff
@@ -29,10 +28,12 @@ const FILE_PATH: &str = "/dev/urandom";
 const FILE_PATH: &str = "/dev/random";
 
 pub fn getrandom_inner(dest: &mut [u8]) -> Result<(), Error> {
-    lazy_static! {
-        static ref FILE: Result<File, Error> = init_file();
-    }
-    let mut f = FILE.as_ref()?;
+    static ONCE: Once = Once::new();
+    static mut FILE: Option<Result<File, Error>> = None;
```

Comment:
Why not return to storing file descriptor in `AtomicIsize` like I was doing in #14? AFAIK negative values are invalid for all supported targets, so we could use the highest bit as a flag "initialization is not finished" and the lowest bit will be "initialization is ongoing".

Reply:
Given that we shouldn't be depending on `std` at all. You're correct, we would need to do something like that. Let me take another look.

Reply:
Or you just use an `AtomicBool` for the synchronization, then you don't have to do bit fiddling. ;)

Reply:
However in that case you have multiple shared vars and can't use relaxed loads, so is it worth it?

Reply:
Also, how do you plan to open a file without depending on `std`?
(Seems like a separate issue though from not using `std` for synchronization.)

Reply:
> However in that case you have multiple shared vars and can't use relaxed loads, so is it worth it?

Not sure what you mean? `AtomicBool` doesn't work, I'm sorry, you need three states (unless your initialziation is idempotent, which it probably is not when file descriptiors are involved).

Then you only need atomic accesses (release/acquire) for the atomic flag, everything else can be non-atomic.

Reply:
So, roughly, the algorithm is something like
```rust
const UNINIT: u8 = 0;
const PENDING: u8 = 1;
const DONE: u8 = 2;
static FLAG: AtomicU8 = AtomicU8::new(UNINIT);
static mut DATA: T := <whatever>;

fn init_data() -> T {
  // the lazy init code.
}

fn get_data() -> T {
  loop {
    let flag = FLAG.load(Ordering::Acquire);
    if flag == DONE {
      // Safe because: We know the data is initialized and we did an acquire so
      // we synchronized with the initialization.
      return unsafe { DATA }
    }
    // Let's try to initialize.
    // This could likely be "Relaxed", but let's play safe and use AcqRel.
    if flag == UNINIT && FLAG.compare_and_swap(UNINIT, PENDING, Ordering::AcqRel) == UNINIT {
      // We won the race to initialize!
      let data = init_data();
      // Safe because: we "grabbed the lock" by successfully CASing from UNINIT to PENDING.
      unsafe { DATA = data; }
      // Now signal everyone else that the DATA is ready. Must be "release";
      // this synchronizes with the "acquire" load above.
      FLAG.store(DONE, Ordering::Release);
      return data;
    }
    // Just try again, spinning until we see DONE.
  }
}
```

Reply:
Yes, we have 3 states, so single `AtomicBool` will not work. Also the idea was to store initialization state and file descriptor in a single atomic, so code will use just a single memory load.

BTW if we are aiming to drop `std` we can not use `Once`, so we have to return to custom spinlock from #14. To prevent wastefully burning CPU we can use `libc::nanosleep`.

Reply:
Alternatively, you could look into using https://docs.rs/spin/0.5.0/spin/struct.Once.html directly? You know what code you are going to put into that initializer, so you know if it's fine to spin-wait on that. Seems better than open-coding the same thing. ;)

However, you are putting file system code into there, right? To open `/dev/random` or so? (a) how does that make any sense with nostd and (b) that's really not something you want to do with a spin lock. You might be waiting for spinning metal here. That takes *ages*.

Reply:
Nothing prevents us from using system capabilities in a `no_std` crate via `libc`.

Reply:
> Also the idea was to store initialization state and file descriptor in a single atomic, so code will use just a single memory load.

Does this actually help with performance, or does it just make the code more complicated to think about by mixing up (synchronization) control and data?

> BTW if we are aiming to drop std we can not use Once, so we have to return to custom spinlock from #14. To prevent wastefull burning of CPU we can use libc::nanosleep.

Another one of these cyclic dependencies with libstd. :/ `parking_lot` has that problem, too.

Reply:
> Nothing prevents us from using system capabilities in a no_std crate via libc.

It's super annoying though to not be able to use `File`. :/

Reply:
On a second thought keeping all state in a single atomic will not work, `fetch_or` in the initialization loop may corrupt file descriptor value. So we return to this [code](https://github.com/newpavlov/getrandom/blob/6ab631c27646b815beb709c6f7b196563094e5cc/src/use_file.rs) (but we can relax some orderings, see comments in the PR). But I think it's worth to keep `-1` as an initialization value for file descriptor atomic to prevent wasteful memory reads.

>It's super annoying though to not be able to use File

Yeah, but luckily we only need `open` and `read`, though duplicating read loop will be indeed annoying.

Reply:
The example @RalfJung posted above uses more precise synchronisation ([this article](https://preshing.com/20120913/acquire-and-release-semantics/) gives a better description of the locks required). Though as noted we will want some type of `thread::yield` or sleep when waiting.

Reply:
> The example @RalfJung posted above uses more precise synchronisation ([this article](https://preshing.com/20120913/acquire-and-release-semantics/) gives a better description of the locks required). Though as noted we will want some type of `thread::yield` or sleep when waiting.

[`libc::sched_yield`](https://docs.rs/libc/0.2.58/libc/fn.sched_yield.html) is what we want here.

> It's super annoying though to not be able to use File. :/

Agree 100%. However, we only need to do this complex stuff (both opening files, and non-trivial syncronization) on `unix` targets, so we can just use `libc::open`, `libc::read`, and `libc::close`. It will be like writing C code (not exactly a good thing, but it will work).

> On a second thought keeping all state in a single atomic will not work

So I think I have an idea that will allow us to keep using a single value, and not make the code too terrible, I'll update this CL with my proposal.

Reply:
>libc::sched_yield is what we want here.

Don't forget about rust-lang/rust#46774. We have discussed it in your initial lazy_static PR, I think sleeping short interval will be a better option in our case.

Reply:
AFAIK `parking_lot` has some fancy back-off logic that would, in principle, be useful here. @Amanieu is that available as a separate crate or so?

However, this is also not very performance sensitive, as it can only happen a few times and will never happen once initialization is complete, so `sleep_ms(1)` is likely good enough.

Reply:
> Don't forget about [rust-lang/rust#46774](https://github.com/rust-lang/rust/issues/46774).

Whoops I had forgotten, and sleeping for 1 ms sounds like the best bet. Ideally (at least in my opinion) `parking_lot` would just be the crate that implements all the synchronization primitives for  the stdlib, making this all a lot easier. But until then, I think we will have to roll our own less good versions.

Reply:
It [will be](https://github.com/rust-lang/rust/pull/56410). But even then it will not be possible for a `no_std` crate to use it as there is a cyclic dependency between libstd and parking_lot.

I suggest we don't add `getrandom` to that cycle. Things are complicated enough already.^^